### PR TITLE
Added support for the STM32F429I-DISCO kit by STMicro.

### PR DESCRIPTION
**Pull request by tobbad**
Sorry for opening another pull request. For comment on the last request see:
https://github.com/micropython/micropython/pull/1641
Merging the changes on upstream/master was not as easy as I thought...
# Intent

With this pull request I would like to add minimal support for the STM32F429I-Discovery kit of STM. USB-CDC and USB-MSC interface works but there is no support for the on board sensors and the LCD. 
# Changes in stmhal subfolder
- **flash.c**: Added definitions for stm32f429 which has slightly more flash. Added calculation of flash table directly from table instead using a hard coded value and there by remove a magic constant. 
- **storage.c**: Added defintition for stm32f429 as copy of STM32F405xx configuration.
- **usbd_conf.c**: Added USB setup for stm32f429 as copy of stm32f405/7 but on the USB_OTG_HS instance. The formatting of the code does not follow micropythons C Code conventions with respect to braces. I just did the minimal changes. 
- **boards/stm32f429_af.csv**: Alternate pin function defines for stm32f429.
- **boards/stm32f429.ld**: Linkerscript for stm32f429.
- **boards/STM32F429DISC/**: Board definion files for STM32F429i-Disco kit.
- **uart.c** Added a default initialization as the variable `GPIO_Pin2` is no more touched in the STM32F429IDISCO kit
- **stm32_it.c/h**: Using `USE_USB_HS_IN_FS` to select code based on the configuration in stm32fxx_hal_conf. Using `USE_USB_HS_IN_FS`  enables the embedded FS phy in the HS part of the chip.
# Comment to realized changes with respect to last pull request
- **`pins.csv`**: All pins are now included. The sequence P1 pin1 to P2 pin 64.
- **`flash.c`**: Added #define for table size using the `MP_ARRAY_SIZE` macro.
- **`usbd_conf.c`**: Going back to Windows line ending style (as in micropython/master). USING new #define `USE_USB_HS_IN_FS` to activate FS code for HS part of device.

### CODE REVIEW

File: stmhal/usbd_conf.c [hunk 1]
```diff
@@ -96,9 +96,51 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
...
+    /* Configure DM DP Pins */
+    GPIO_InitStruct.Pin = (GPIO_PIN_14 | GPIO_PIN_15);
+    GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Speed = GPIO_SPEED_HIGH;
+    GPIO_InitStruct.Alternate = GPIO_AF12_OTG_HS_FS;
+    HAL_GPIO_Init(GPIOB, &GPIO_InitStruct);
+
+#if defined(MICROPY_HW_USB_VBUS_DETECT_PIN)
+    /* Configure VBUS Pin */
```
Comment (neilh10) [inline stmhal/usbd_conf.c]: The USB_HS VBUS & OTG pins are dedicated and optional per USB hw instance. 
For a simple  USB_HS HOST on  pybv10 these need to  be separate from USB_FS, and could be
MICROPY_HW_USB_HS_VBUS_DETECT_PIN  
MICROPY_HW_USB_HS_OTG_DETECT_PIN
(Sorry for not suggesting it on earlier comment)

Comment (tobbad) [inline stmhal/usbd_conf.c]: I think you make a valid suggestion. But I would like to keep my changes to the common code very small and related to the topic (Added support for the STM32F429I-DISCO kit by STMicro.). Therefore I would like to keep it as it is for the merge to the main. As the current state of micropython does not support 2 USB connection I would like to keep it as proposed.

If you later make a pull request to support 2 Phys you probably have to solve the problem as 
you propose.
@dhylands What do you think?

Comment (dhylands) [inline stmhal/usbd_conf.c]: Yeah - I think that adding support for simultaneous HS/FS phys really has nothing to do with STM32F4DISCO per-say, and should be done in a separate PR.

Comment (neilh10) [inline stmhal/usbd_conf.c]: Sure no problem. If I make the update later to the common code  stmhal/usbd_conf.c I think it works for the STM32F429I as well.

File: stmhal/flash.c [hunk 2]
```diff
@@ -83,12 +101,27 @@ static const uint32_t flash_info_table[26] = {
     ADDR_FLASH_SECTOR_10, FLASH_SECTOR_10,
     ADDR_FLASH_SECTOR_11, FLASH_SECTOR_11,
     #endif
+	#if defined(FLASH_SECTOR_12)
```
Comment (dpgeorge) [inline stmhal/flash.c]: Please use spaces not tabs.

File: stmhal/flash.c [hunk 3]
```diff
@@ -83,12 +101,27 @@ static const uint32_t flash_info_table[26] = {
...
+	ADDR_FLASH_SECTOR_23, FLASH_SECTOR_23,
+	#endif
     ADDR_FLASH_END, 0,
 };
+#define FLASH_INFO_TABLE_SIZE MP_ARRAY_SIZE(flash_info_table)
 
 uint32_t flash_get_sector_info(uint32_t addr, uint32_t *start_addr, uint32_t *size) {
     if (addr >= flash_info_table[0]) {
-        for (int i = 0; i < 24; i += 2) {
+        for (int i = 0; i < FLASH_INFO_TABLE_SIZE-2; i += 2) {
```
Comment (dpgeorge) [inline stmhal/flash.c]: No need to define FLASH_INFO_TABLE_SIZE, just write MP_ARRAY_SIZE(flash_info_table) directly (also spaces around -).

File: stmhal/uart.c [hunk 4]
```diff
@@ -115,7 +115,7 @@ void uart_deinit(void) {
 STATIC bool uart_init2(pyb_uart_obj_t *uart_obj) {
     USART_TypeDef *UARTx;
     IRQn_Type irqn;
-    uint32_t GPIO_Pin, GPIO_Pin2;
+    uint32_t GPIO_Pin, GPIO_Pin2= NULL;
```
Comment (dpgeorge) [inline stmhal/uart.c]: I don't understand why this is needed.  It builds cleanly without this change.

File: stmhal/boards/stm32f429.ld [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+/* Specify the memory areas */
+MEMORY
+{                                                   
+    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 0x0200000 /* entire flash, 2048 KiB */
+    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 0x0004000 /* sector 0, 16 KiB */
+    FLASH_TEXT (rx) : ORIGIN = 0x08020000, LENGTH = 0x088000 /* sectors 5,6,7,8, 4*128KiB = 512 KiB (could increase it more) */
+    FLASH_FAT (r)   : ORIGIN = 0x08100000, LENGTH = 0x0100000 /* sectors 12-23 = 1MByte */
+    RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 0x0020000 /* 128 KiB */
+    SDRAM(xrw)      : ORIGIN = 0xC0000000, LENGTH = 0x0800000 /* 8 MByte */
```
Comment (dpgeorge) [inline stmhal/boards/stm32f429.ld]: Does it have 64mbyte or 64m bit SDRAM?

File: stmhal/boards/stm32f429.ld [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+*/
+
+/* Specify the memory areas */
+MEMORY
+{                                                   
+    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 0x0200000 /* entire flash, 2048 KiB */
+    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 0x0004000 /* sector 0, 16 KiB */
+    FLASH_TEXT (rx) : ORIGIN = 0x08020000, LENGTH = 0x088000 /* sectors 5,6,7,8, 4*128KiB = 512 KiB (could increase it more) */
+    FLASH_FAT (r)   : ORIGIN = 0x08100000, LENGTH = 0x0100000 /* sectors 12-23 = 1MByte */
+    RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 0x0020000 /* 128 KiB */
```
Comment (dpgeorge) [inline stmhal/boards/stm32f429.ld]: Doesn't it have 256k SRAM?

File: stmhal/boards/stm32f429.ld [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    GNU linker script for STM32F429i-Discovery kit with external 64MByte SDRam
+*/
+
+/* Specify the memory areas */
+MEMORY
+{                                                   
+    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 0x0200000 /* entire flash, 2048 KiB */
+    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 0x0004000 /* sector 0, 16 KiB */
+    FLASH_TEXT (rx) : ORIGIN = 0x08020000, LENGTH = 0x088000 /* sectors 5,6,7,8, 4*128KiB = 512 KiB (could increase it more) */
+    FLASH_FAT (r)   : ORIGIN = 0x08100000, LENGTH = 0x0100000 /* sectors 12-23 = 1MByte */
```
Comment (dpgeorge) [inline stmhal/boards/stm32f429.ld]: Is this used?

File: stmhal/storage.c [hunk 8]
```diff
@@ -56,6 +56,13 @@ STATIC byte flash_cache_mem[0x4000] __attribute__((aligned(4))); // 16k
 #define FLASH_MEM_SEG1_START_ADDR (0x08004000) // sector 1
 #define FLASH_MEM_SEG1_NUM_BLOCKS (128) // sectors 1,2,3,4: 16k+16k+16k+16k(of 64k)=64k
 
+#elif defined(STM32F429xx)
+
+#define CACHE_MEM_START_ADDR (0x10000000) // CCM data RAM, 64k
+#define FLASH_SECTOR_SIZE_MAX (0x10000) // 64k max, size of CCM
+#define FLASH_MEM_SEG1_START_ADDR (0x08004000) // sector 1
+#define FLASH_MEM_SEG1_NUM_BLOCKS (224) // sectors 1,2,3,4: 16k+16k+16k+64k=112k
```
Comment (dpgeorge) [inline stmhal/storage.c]: If you don't mind wasting half the flash, you can start this at sector 12 and make it use the last 1mb of flash (which will be 512k for the fatfs due to the 64k cache size).  You could also use the SDRAM for the cache and then utilise all the 1mb of flash for the fatfs.

File: stmhal/flash.c [hunk 9]
```diff
@@ -27,6 +27,8 @@
 #include STM32_HAL_H
 
 #include "flash.h"
+#include "mpconfigport.h"
+#include "../py/misc.h"
```
Comment (dpgeorge) [inline stmhal/flash.c]: This can be "py/misc.h".

File: stmhal/usbd_conf.c [hunk 10]
```diff
@@ -60,15 +60,15 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
   {
     /* Configure USB FS GPIOs */
     __GPIOA_CLK_ENABLE();
-    
+
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: Please, no cosmetic changes in this PR, just functional ones related to the F429.

File: stmhal/usbd_conf.c [hunk 11]
```diff
@@ -60,15 +60,15 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
...
     GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
     GPIO_InitStruct.Pull = GPIO_NOPULL;
     GPIO_InitStruct.Speed = GPIO_SPEED_HIGH;
     GPIO_InitStruct.Alternate = GPIO_AF10_OTG_FS;
-    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct); 
-    
-	/* Configure VBUS Pin */
+    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
+
+    /* Configure VBUS Pin */
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: Cosmetic.

File: stmhal/usbd_conf.c [hunk 12]
```diff
@@ -76,14 +76,14 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
     GPIO_InitStruct.Pull = GPIO_NOPULL;
     HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
 #endif
-	
+
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: Cosmetic.

File: stmhal/usbd_conf.c [hunk 13]
```diff
@@ -76,14 +76,14 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
...
-	
+
 #if defined(MICROPY_HW_USB_OTG_ID_PIN)
     // USB ID pin is always A10
     GPIO_InitStruct.Pin = GPIO_PIN_10;
     GPIO_InitStruct.Mode = GPIO_MODE_AF_OD;
     GPIO_InitStruct.Pull = GPIO_PULLUP;
     GPIO_InitStruct.Alternate = GPIO_AF10_OTG_FS;
-    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct); 
+    HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: Cosmetic.

File: stmhal/usbd_conf.c [hunk 14]
```diff
@@ -150,10 +192,10 @@ void HAL_PCD_MspInit(PCD_HandleTypeDef *hpcd)
     GPIO_InitStruct.Pull = GPIO_NOPULL;
     GPIO_InitStruct.Alternate = GPIO_AF10_OTG_HS;
     HAL_GPIO_Init(GPIOI, &GPIO_InitStruct);
-
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: Cosmetic.

File: stmhal/usbd_conf.c [hunk 15]
```diff
@@ -252,8 +293,8 @@ void HAL_PCD_ResetCallback(PCD_HandleTypeDef *hpcd)
   case PCD_SPEED_FULL:
     speed = USBD_SPEED_FULL;    
     break;
-	
-	default:
+    
+    default:
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: Cosmetic.

File: stmhal/usbd_conf.c [hunk 16]
```diff
@@ -362,10 +403,36 @@ USBD_StatusTypeDef  USBD_LL_Init (USBD_HandleTypeDef *pdev)
   HAL_PCD_SetTxFiFo(&pcd_handle, 1, 0x40);
   HAL_PCD_SetTxFiFo(&pcd_handle, 2, 0x20);
   HAL_PCD_SetTxFiFo(&pcd_handle, 3, 0x40);
+#endif
 
+#if defined(USE_USB_HS_IN_FS)
```
Comment (dpgeorge) [inline stmhal/usbd_conf.c]: I think this and the #endif above should be a #elif.

File: stmhal/uart.c (hunk 4 above)
Comment (tobbad) [inline stmhal/uart.c]: Enter
arm-none-eabi-gcc -I../lib/netutils -I. -I.. -Ibuild-STM32F429DISC -Icmsis/inc -Icmsis/devinc -Ihal/f4/inc -Iusbdev/core/inc -Iusbdev/class/inc -I../lib/mp-readline -I../lib/netutils -I../lib/timeutils -Wall -Wpointer-arith -Werror -ansi -std=gnu99 -nostdlib  -DSTM32F429xx -mthumb -mabi=aapcs-linux -mfpu=fpv4-sp-d16 -mfloat-abi=hard -fsingle-precision-constant -Wdouble-promotion -mtune=cortex-m4 -mcpu=cortex-m4 -DMCU_SERIES_F4 -O0 -Iboards/STM32F429DISC -DSTM32_HAL_H='<stm32f4xx_hal.h>' -g -DPENDSV_DEBUG -c -MD -o build-STM32F429DISC/uart.o uart.c
Gives me an error of uninitialized variable which is true if you check the code. I'll check why the makefile does not complain.

Comment (dhylands) [inline stmhal/uart.c]: make DEBUG=1 causes the same error (even for the pyboard).

I'm going to do a PR just for that change, and add DEBUG=1 builds to my set of travis builds.

Comment (tobbad) [inline stmhal/uart.c]: It is the -O0 option in DEBUG whereas in Release -Os is used.

File: stmhal/boards/stm32f429.ld (hunk 5 above)
Comment (tobbad) [inline stmhal/boards/stm32f429.ld]: SDRAM is 64m Bit = 8MByte. i'll fix the comment.

File: stmhal/boards/stm32f429.ld (hunk 6 above)
Comment (tobbad) [inline stmhal/boards/stm32f429.ld]: Yes. I'll fix it.

File: stmhal/boards/stm32f429.ld (hunk 7 above)
Comment (tobbad) [inline stmhal/boards/stm32f429.ld]: Removed.

File: stmhal/storage.c (hunk 8 above)
Comment (tobbad) [inline stmhal/storage.c]: By purpose I keep the port very close to the 407 version. In a later PR I'll changet the linkerscript and the storage layout.

File: stmhal/usbd_conf.c (hunk 16 above)
Comment (tobbad) [inline stmhal/usbd_conf.c]: Yes.

File: stmhal/boards/stm32f429.ld [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    GNU linker script for STM32F429i-Discovery kit with external 8MByte SDRam
+*/
+
+/* Specify the memory areas */
+MEMORY
+{                                                   
+    FLASH (rx)      : ORIGIN = 0x08000000, LENGTH = 0x0200000 /* entire flash, 2048 KiB */
+    FLASH_ISR (rx)  : ORIGIN = 0x08000000, LENGTH = 0x0004000 /* sector 0, 16 KiB */
+    FLASH_TEXT (rx) : ORIGIN = 0x08020000, LENGTH = 0x0088000 /* sectors 5,6,7,8, 4*128KiB = 512 KiB (could increase it more) */
+    RAM (xrw)       : ORIGIN = 0x20000000, LENGTH = 0x0030000 /* 192 KiB */
```
Comment (tobbad) [inline stmhal/boards/stm32f429.ld]: The stm32f429 has 256 kByte RAM but only 192 are at 0x20000000 the remaining 64kByte is at 0x10000000 (CCM).

**Comment by tobbad** [comment]
I updated the code according to the review input.

**Comment by dpgeorge** [comment]
Rebased and merged in 8844d031e4bc39cb6e6cd053b63a8cdaeb383221.

Great work, thanks!  Good that you found the solution to the wfi problem.
